File: Depth-First Search/Binary tree/LET_no606_Construct String from Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def tree2str(self, root: Optional[TreeNode]) -> str:
        res = []
        self.dfs(root, res)
        return ''.join(res)
    
    def dfs(self, cur, res):
        if not cur:
            return
        
        res.append(str(cur.val))
        if not cur.left and not cur.right:
            return
        
        res.append('(')
        self.dfs(cur.left, res)
        res.append(")")
        
        if cur.right:
            res.append('(')
            self.dfs(cur.right, res)
            res.append(')')

File: Depth-First Search/Binary tree/LET_no606_Construct String from Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    string tree2str(TreeNode* root) {
        string res;
        dfs(root, res);
        return res;
    }

    void dfs(TreeNode* cur, string& res) {
        if (!cur)
            return;
            
        res.append(to_string(cur->val));
        if (!cur->left && !cur->right)
            return;
        
        res.push_back('(');
        dfs(cur->left, res);
        res.push_back(')');

        if (cur->right) {
            res.push_back('(');
            dfs(cur->right, res);
            res.push_back(')');
        }
    }
};